### PR TITLE
WT-11877 Add configuration for maximum wait time between fsync calls

### PR DESCRIPTION
**Pull request by y123456yz**
for example: we configure io_capacity.tatal=1M, where fsync background takes 10 seconds to work, which is too long.
Also, if the io_capacity.tatal‘s value is large(1T), sync background should takes a long time to work. If there is a write operation, but the threshold (cap->written < cap->threshold) is not met for a long time, shold we force sync background ?

### CODE REVIEW

**Comment by y123456yz** [comment]
also, I have a question, why does capacity work thread only work when io_capacity.tatal>0 is configured, and if so, what is the point of _capacity server threads?

If io_capacity is not configed, can it be enabled for background fsync?

thanks.

**Review by ddanderson — COMMENTED** [review]
@y123456yz, it would be nice to have a test associated with this change, showing it does what it says it does.  Would you be willing to add a test?

If so, tests are typically in Python (see test/suite/test_*.py for many examples).  What I have in mind is setting the cache size to be small, set `fsync_background_period_sec` to a small number, like `1`, and add enough data to a table to cause some eviction, make sure at least 1 second has elapsed, and check the fsync statistic to see that we've done some fsyncing.  I know our current tests for capacity just check that configuring it is possible, but it would be good to have something stronger to support this change.

**Comment by y123456yz** [comment]
> @y123456yz, it would be nice to have a test associated with this change, showing it does what it says it does. Would you be willing to add a test?
> 
> If so, tests are typically in Python (see test/suite/test_*.py for many examples). What I have in mind is setting the cache size to be small, set `fsync_background_period_sec` to a small number, like `1`, and add enough data to a table to cause some eviction, make sure at least 1 second has elapsed, and check the fsync statistic to see that we've done some fsyncing. I know our current tests for capacity just check that configuring it is possible, but it would be good to have something stronger to support this change.

hi, @ddanderson 
got it, I have a lot of other things to do these days, so it may take some time. After writing the test code, I will @ you, expected to be completed the day after tomorrow.

thanks.

**Comment by ddanderson** [comment]
@y123456yz, thank you!  Can you also merge develop into your branch, as there are some new conflicts?  There was some recent work put into develop to use io_capacity for the "chunk cache", but I think it should be compatible with your work.

**Comment by y123456yz** [comment]
@ddanderson @korteland 
I have add the python unit test, please review the code again，thanks.

File: dist/api_data.py
Comment (ddanderson) [inline dist/api_data.py:744]: Can you change the wording a bit (and fix spelling mistake in the name).  I suggest:
```
Config('fsync_background_period_sec', '0', r'''
max waiting period for background fsync. If the written threshold is not met in this time, a background fsync is done.''',
```

File: src/conn/conn_capacity.c
Comment (ddanderson) [inline src/conn/conn_capacity.c:46]: Spelling: backgroud should be "background".  Please fix globally.

File: src/include/capacity.h
Comment (ddanderson) [inline src/include/capacity.h:17]: Spelling: "fsync".  Please change globally.

**Review by ddanderson — COMMENTED** [review]
@y123456yz nice work with the test.  There are spelling and formatting things to fix.  Otherwise, would be approved.

**Comment by ddanderson** [comment]
@y123456yz, in addition to the review comments, I see that two tests for windows failed.  One fails this way:
```
2023/12/04 11:11:18.032]   File "C:\data\mci\3e5d411e22d6f8fea7370b5bff09e44d\wiredtiger\test\suite\test_io_capacity01.py", line 121, in test_io_catacity_fsync_backgroud_period
[2023/12/04 11:11:18.032]     self.assertGreater(self.get_stat(stat.conn.fsync_all_fh_total), 0)
[2023/12/04 11:11:18.032] AssertionError: 0 not greater than 0
```
I don't know why this test behaves differently on Windows.  The statistic is incremented in the common OS code, before the native windows API is called.  And in fact, it appears to me that `__fsync_background` will end up doing nothing on Windows.  But the statistic should get incremented.  Maybe eviction behaves differently in relation to checkpoint on Windows.

The other fault is a crash.  At the moment, that's all we have without pursuing it on a Windows box.
```
[2023/12/04 10:47:41.223] [pid:3956]: None ... [pid:3956]: test_io_capacity01.test_io_catacity_01.test_io_catacity_written_threshold: starting
[2023/12/04 10:47:44.678] fi
[2023/12/04 10:47:44.678] bash: line 44:   972 Segmentation fault      /cygdrive/c/python/Python311/python ../test/suite/run.py -v 2 -b 7/12 $threads_command 2>&1
```
Let me know if you have any ideas on these?

**Comment by y123456yz** [comment]
> Let me know if you have any ideas on these?

hi, @ddanderson 
Sorry, I'm linux tech stack, I don't understand windows tech stack. Can you Invite someone who knows windows to look at this?
thanks.

**Comment by y123456yz** [comment]
@ddanderson 
I have already solved this problem, please review again.
sorry, The reason for the problem is that a word was typed incorrectly。

**Comment by y123456yz** [comment]
@korteland
please help review this PR
thanks.
:)

**Comment by korteland** [comment]
@y123456yz sorry for the delay here - I'm playing second fiddle to Don in this review, so I was waiting for his review to roll in. I'll take a look now.

File: src/conn/conn_api.c
Comment (korteland) [inline src/conn/conn_api.c:1817]: The changes in this file are unnecessary.

File: dist/api_data.py
Comment (korteland) [inline dist/api_data.py:749]: Can you change the stat name to `fsync_maximum_wait_period` please, both here and in other parts of the code? I think that would be a closer match to the behaviour.

File: src/conn/conn_capacity.c
Comment (korteland) [inline src/conn/conn_capacity.c:39]: I think there's a merge that went wrong here - there's already some checks below for `io_capacity.total`.

File: src/include/capacity.h
Comment (korteland) [inline src/include/capacity.h:17]: This macro should have a `MAX` in the name somewhere so the purpose is clearer.

File: src/include/capacity.h
Comment (korteland) [inline src/include/capacity.h:58]: This comment could use some improvements - we can get away with something shorter since the full description is in the config. The below is a suggestion, but feel free to do your own thing.

`Maximum period between fsync calls`

File: test/evergreen.yml
Comment (korteland) [inline test/evergreen.yml:3038]: Typo: `catacity`

File: src/conn/conn_capacity.c
Comment (korteland) [inline src/conn/conn_capacity.c:48]: Spelling/grammar: "exceed maxmum" -> "exceeds maximum"

File: src/conn/conn_capacity.c
Comment (korteland) [inline src/conn/conn_capacity.c:137]: These can be on the same line

File: src/conn/conn_capacity.c
Comment (korteland) [inline src/conn/conn_capacity.c:104]: Could you explain the logic of this change? I don't completely get it.

File: test/suite/test_io_capacity01.py
Comment (korteland) [inline test/suite/test_io_capacity01.py:1]: I won't get time today to review the test sorry - I'll have to do this later (or once the other comments are sorted out).

File: src/conn/conn_capacity.c
Comment (y123456yz) [reply src/conn/conn_capacity.c:104]: mainly to avoid a long time without triggering fsyn. after we add fsync_background_period_sec Feature, there is no need to change here.

File: test/suite/test_io_capacity01.py
Comment (y123456yz) [reply test/suite/test_io_capacity01.py:1]: got it, thanks.

File: src/conn/conn_capacity.c
Comment (korteland) [reply src/conn/conn_capacity.c]: This is still an issue - "exceed" needs to be "exceeds".

File: src/conn/conn_capacity.c
Comment (korteland) [reply src/conn/conn_capacity.c]: Cool - thanks for explaining that.

File: test/suite/test_io_capacity01.py
Comment (korteland) [inline test/suite/test_io_capacity01.py:119]: It would be more robust to spin on the statistic rather than sleeping.

Comment (y123456yz) [reply test/suite/test_io_capacity01.py:119]: hi, @korteland 
how to spin, please guide me, I haven't used this before. To be honest, I'm not very familiar with Python.
thanks.

Comment (korteland) [reply test/suite/test_io_capacity01.py]: Hi @y123456yz , no problem. So rather than sleeping, I would have some code like this:

```
while (self.get_stat(stat.conn.capacity_bytes_written) == 0):
    attempts += 1
    time.sleep(0.1)
    if attempts > 10000:
        throw Exception("Timed out waiting for statistic")
```

The numbers are all made up, and you should clean it up and experiment to see what works. But this is the overall idea.

Comment (y123456yz) [reply test/suite/test_io_capacity01.py]: @korteland 
got it, thanks.
:)

**Comment by korteland** [comment]
@y123456yz thanks for the cleanup - still a couple of unresolved threads above.

**Comment by korteland** [comment]
@y123456yz what's the status of this? There are still a couple of unresolved issues that have been open for a while.

File: test/suite/test_io_capacity01.py
Comment (y123456yz) [reply test/suite/test_io_capacity01.py:80]: @korteland 
thanks, I have remove the capacity_bytes_written.

**Comment by y123456yz** [comment]
@korteland 
I'm sorry to forget that, please review the code again.
thanks.
:)

File: src/conn/conn_api.c
Comment (korteland) [reply src/conn/conn_api.c:1]: @y123456yz This is still unaddressed.

File: test/suite/test_io_capacity01.py
Comment (korteland) [inline test/suite/test_io_capacity01.py:42]: Nitpick: this comment is still not quite formatted correctly.

**Review by korteland — COMMENTED** [review]
Thanks - almost there, just whitespace changes.

**Comment by y123456yz** [comment]
@korteland 
Thank you for your patience, I was so careless, please review again.

File: src/conn/conn_api.c
Comment (korteland) [reply src/conn/conn_api.c:1]: @y123456yz Please address this comment.

**Review by korteland — APPROVED** [review]
Thanks for your patience @y123456yz - LGTM! I'm going to request another review from Don to give him a last chance to (n)ack.

File: src/conn/conn_api.c
Comment (y123456yz) [reply src/conn/conn_api.c:1]: done, thanks.

File: src/conn/conn_capacity.c
Comment (ddanderson) [inline src/conn/conn_capacity.c:51]: There is already a check for over 60, it is enforced because of the new code in api_data.py, where `max='60'`.  So I would change this to merely:
```
    if (cval.val != 0)
       cap->fsync_maximum_wait_period = (uint64_t)cval.val;
```

I wanted to point out though, that the construct in the PR at line 47 can be unsafe.  That is, assume there was not the upper and lower limit check in api_data.py.  Then the caller could have this in the config string:
```
    io_capacity=(fsync_maximum_wait_period=-10)
```
cval.val is a signed value, so the `if` clause would _not_ be taken, no error issued and the line `cap->fsync_maximum_wait_period = (uint64_t)cval.val;` would get executed (instead of issuing the error).  I would avoid mistakes like this by using this idiom:
```
    if (cval.val != 0) {
        ...any other checks...
        cap->.... = (uint64_t)cval.val;
   }
```
It would be good to use this idiom whenever checking config strings.

File: src/include/capacity.h
Comment (ddanderson) [inline src/include/capacity.h:18]: I don't think this define is needed, as the max is in api_data.py

File: src/conn/conn_capacity.c
Comment (ddanderson) [inline src/conn/conn_capacity.c:41]: Use idiom:
```
   if (cval.val != 0) {
        ...any other checks...
        cap->.... = (uint64_t)cval.val;
   }
```
as mentioned in previous comment.  It's not unsafe here, but it's good to always use a safe idiom.

**Review by ddanderson — COMMENTED** [review]
I tagged something that could be simplified/improved.

We still need to figure out why we have failures on Windows before we can merge.

**Comment by y123456yz** [comment]
hi, @ddanderson 
Follow your instructions, I have change the code. 
please review again.

The reason for the failure on windos is that I used “fsync_background_period_sec” in the script api_data.py stat_data.py and test_io_capacity01,  but "fsync_backgroud_period_secis" used in the function __capacity_config. you can see:
https://github.com/wiredtiger/wiredtiger/pull/9797/commits/0525c38dad4a435abd16292bb3def55d75e2d84a

In fact, there may also be issues on the Linux platform, and running ctest on the Linux platform may also cause exceptions.
thanks.

**Comment by ddanderson** [comment]
The reason the windows tests are failing is this code in conn_capacity.c
```
    if (F_ISSET(conn, WT_CONN_IN_MEMORY | WT_CONN_READONLY) || !__wt_fsync_background_chk(session))
        return (0);
```
The `__wt_fsync_background_chk` sees that WT has no code to support a background fsync on Windows (it may or may not be possible to do, but in any case WT does not currently support it).

And because of that, the capacity server doesn't run on Windows, so that statistics expected by the tests are never incremented.

**Comment by y123456yz** [comment]
@ddanderson 
Following your guidance, I have modified the code and rerun ./s_all, please review again.
thanks.

**Comment by y123456yz** [comment]
@ddanderson 
Following your guidance, I have modified the code, please review again.

**Comment by korteland** [comment]
@y123456yz there are still quite a number of failures in the PR testing. They seem to be an opaque `AttributeError: 'ModuleImportFailure' object has no attribute 'simpleName'`, and I'm not sure what's causing it. I would suggest three things:
- Merging master into your branch
- Running the Python tests
- Running `s_all`

Hopefully that's useful - let me know how you go.

**Comment by y123456yz** [comment]
hi, @korteland 

[root@localhost dist]# ./s_all
dist/s_all run started...
######### s_all run of: s_string resulted in:
s_string skipped: aspell not found
#########
######### s_all run of: s_docs resulted in:
s_docs skipped: unsupported version of doxygen: 1.8.5, not 1.8.17, 1.9.1, or 1.9.3
#########
######### s_all run of: s_evergreen resulted in:
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
s_evergreen: /root/yyz/wiredtiger-yyz-master/wiredtiger/test/evergreen/evg_cfg.py check
ERROR [/root/yyz/wiredtiger-yyz-master/wiredtiger/test/evergreen/evg_cfg.py]: need to run this script inside a wiredtiger repo
        `git config remote.origin.url` returned "https://xxx@github.com/y123456yz/wiredtiger"
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
#########
######### s_all run of: comment_style.py resulted in:
Illegal comment in test/cppsuite/src/common/constants.h:115 } // namespace test_harness
Illegal comment in test/unittest/tests/utils.h:13 } // namespace utils
Detected 2 comment format issues!
#########
dist/s_all run finished. Error? 1
Fatal errors reported by: s_evergreen comment_style.py
[root@localhost dist]# 
[root@localhost dist]# 
[root@localhost dist]# 
[root@localhost dist]# ./s_evergreen
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
./s_evergreen: /root/yyz/wiredtiger-yyz-master/wiredtiger/test/evergreen/evg_cfg.py check
ERROR [/root/yyz/wiredtiger-yyz-master/wiredtiger/test/evergreen/evg_cfg.py]: need to run this script inside a wiredtiger repo
        `git config remote.origin.url` returned "https://xxx@github.com/y123456yz/wiredtiger"
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[root@localhost dist]# 
[root@localhost dist]# python  comment_style.py
Illegal comment in test/cppsuite/src/common/constants.h:115 } // namespace test_harness
Illegal comment in test/unittest/tests/utils.h:13 } // namespace utils
Detected 2 comment format issues!
[root@localhost dist]# 
[root@localhost dist]# 


I have merge the develop branch, run ./s_all again, and Solved some issues.
sorry, due to some software installation failures in my testing environment, such as aspell, So, could you please help me run? 
thanks.

**Comment by y123456yz** [comment]
@korteland 
If it's convenient, please help merge the code. The error in sAll should be my own environment issue.

thanks.
:)

**Comment by korteland** [comment]
@y123456yz I'm still waiting for Don - looking at his calendar, he will likely be available for this next week. Sorry for the delay on this PR!

File: src/conn/conn_capacity.c
Comment (ddanderson) [inline src/conn/conn_capacity.c:53]: I don't think we need this check nor the message, because the config system has already made the check.  Since we have:
```
        Config('fsync_maximum_wait_period', '0', r'''
            max waiting period for background fsync. If the written threshold is not met in this
            time, a background fsync is done.''',
            min='0', max='60'),
```
in `dist/api_data.py`, with a `min` set to 0.

**Review by ddanderson — APPROVED** [review]
I've tagged some code that I think is not needed.  I've marked as "approve", I don't need to re-review if you make the change.

File: src/conn/conn_capacity.c
Comment (y123456yz) [reply src/conn/conn_capacity.c]: I have removed this useless check, thanks.

**Comment by y123456yz** [comment]
@ddanderson @korteland 
Thank you both for your guidance.

**Comment by ddanderson** [comment]
@korteland I'm good with the changes now.
Thank you @y123456yz on your patience through this process.

**Comment by korteland** [comment]
@y123456yz This failed on the new test you added. The error was:

```
ERROR: test_io_capacity01.test_io_capacity01.test_io_capacity_fsync_background_period
----------------------------------------------------------------------
testtools.testresult.real._StringException: Traceback (most recent call last):
  File "/data/mci/5cb621b1178ba2ebb18730af848c0f69/wiredtiger/test/suite/wttest.py", line 233, in _callTestMethod
    method()
  File "/data/mci/5cb621b1178ba2ebb18730af848c0f69/wiredtiger/test/suite/test_io_capacity01.py", line 118, in test_io_capacity_fsync_background_period
    raise Exception("Timed out waiting for fsync_all_fh_total statistic")
Exception: Timed out waiting for fsync_all_fh_total statistic
```

Would you be able to address this please?

**Comment by y123456yz** [comment]
@korteland 
There is no problem running on my Linux testing machine (even if the retry count is 10, there is no problem here). It may be due to insufficient delay time, and I am not sure if this will solve the problem. I have increased the retry cycle from 10 to 11, please help me try again.
Also, does this issue exist in all systems (including Windows and Linux)?
thanks.
:)

**Comment by korteland** [comment]
@y123456yz unfortunately debugging your changes is a little beyond what I have time for. The failure I saw was on Ubuntu 20.04, in an environment where lots of other tests are running. Some general advice I have here is to think carefully about race conditions and failure scenarios - how strong is the guarantee that the test didn't write enough data to trigger an `fsync`?

I'm running now with the timeout of 1.1 seconds, but it may be the case that anything up to 60 seconds is appropriate on an overloaded system.

**Comment by y123456yz** [comment]
hi, @korteland 
Sorry, I didn't consider the actual situation of your testing environment. I don't have a good solution here either Due to the inability of the testing thread and the capacity server thread to collaborate, there is no good solution here.

Or can we increase the timeout time, but I don't know how reasonable it is to set it here? How about 60 seconds?

**Comment by korteland** [comment]
@y123456yz 60 seconds seems reasonable to me - I'm happy to run the tests against it and see if it resolves the issue. If it still fails then I think it's very likely we're looking at an outright bug rather than a spurious scheduling issue.

**Comment by y123456yz** [comment]
@korteland 
I have adjusted the timeout waiting time to 60 seconds，please help run the test In your testing environment.
thanks.

**Comment by korteland** [comment]
@y123456yz that seems to have sorted it out - thanks. I've merged this.

**Pull request merged**
